htlcswitch/iterator.go

```diff
@@ -194,32 +239,141 @@ func (p *OnionProcessor) DecodeHopIterator(r io.Reader, rHash []byte) (HopIterat
 		}
 	}
 
-	return &sphinxHopIterator{
-		nextPacket:      sphinxPacket.NextPacket,
-		processedPacket: sphinxPacket,
-	}, lnwire.CodeNone
+	return makeSphinxHopIterator(onionPkt, sphinxPacket), lnwire.CodeNone
 }
 
-// ExtractErrorEncrypter takes an io.Reader which should contain the onion
-// packet as original received by a forwarding node and creates an
-// ErrorEncrypter instance using the derived shared secret. In the case that en
-// error occurs, a lnwire failure code detailing the parsing failure will be
-// returned.
-func (p *OnionProcessor) ExtractErrorEncrypter(r io.Reader) (ErrorEncrypter, lnwire.FailCode) {
+// DecodeHopIterators performs batched decoding and validation of incoming
+// sphinx packets. For the same `id`, this method will return the same iterators
+// and failcodes upon subsequent invocations.
+//
+// NOTE: In order for the responses to be valid, the caller must guarantee that
+// the presented readers and rhashes *NEVER* deviate across invocations for the
+// same id.
+func (p *OnionProcessor) DecodeHopIterators(id []byte, rs []io.Reader,
+	rHashes [][]byte, incomingCltvs []uint32) ([]HopIterator,
+	[]lnwire.FailCode, error) {
 
-	onionPkt := &sphinx.OnionPacket{}
-	if err := onionPkt.Decode(r); err != nil {
+	batchSize := len(rs)
+
+	var (
+		onionPkts = make([]sphinx.OnionPacket, batchSize)
+		iterators = make([]HopIterator, batchSize)
+		failcodes = make([]lnwire.FailCode, batchSize)
+	)
+
+	tx := p.router.BeginTxn(id, batchSize)
+
+	for i, r := range rs {
+		onionPkt := &onionPkts[i]
+		err := onionPkt.Decode(r)
 		switch err {
+		case nil:
+			// success
```

Comment:
Shouldn't we add a `lnwire.CodeNone` here in the slice above for the fail code? Combined with my suggestion about collapsing some of the arguments and return values into a struct, would seem this way would be easier for the caller to match the failed output to the input that generated it.

Reply:
`lnwire.CodeNone` is the zero value for value codes, so it should already be in that state. If something fails, then it makes the transition to another fail code. I'll add a more descriptive comment to explain this progression